libimage/load.go

```diff
@@ -0,0 +1,79 @@
+package libimage
+
+import (
+	"context"
+	"errors"
+
+	dirTransport "github.com/containers/image/v5/directory"
+	dockerArchiveTransport "github.com/containers/image/v5/docker/archive"
+	ociArchiveTransport "github.com/containers/image/v5/oci/archive"
+	ociTransport "github.com/containers/image/v5/oci/layout"
+)
+
+type LoadOptions struct {
+	CopyOptions
+}
+
+// Load loads one or more images (depending on the transport) from the
+// specified path.  The path may point to an image the following transports:
+// oci, oci-archive, dir, docker-archive.
+func (r *Runtime) Load(ctx context.Context, path string, options *LoadOptions) ([]string, error) {
+	var (
+		loadedImages []string
+		loadError    error
+	)
+
+	if options == nil {
+		options = &LoadOptions{}
+	}
+
+	for _, f := range []func() ([]string, error){
+		// OCI
+		func() ([]string, error) {
+			ref, err := ociTransport.NewReference(path, "")
+			if err != nil {
+				return nil, err
+			}
+			return r.copyFromDefault(ctx, ref, &options.CopyOptions)
+		},
+
+		// OCI-ARCHIVE
+		func() ([]string, error) {
+			ref, err := ociArchiveTransport.NewReference(path, "")
+			if err != nil {
+				return nil, err
+			}
+			return r.copyFromDefault(ctx, ref, &options.CopyOptions)
+		},
+
+		// DIR
+		func() ([]string, error) {
+			ref, err := dirTransport.NewReference(path)
+			if err != nil {
+				return nil, err
+			}
+			return r.copyFromDefault(ctx, ref, &options.CopyOptions)
+		},
+
+		// DOCKER-ARCHIVE
+		func() ([]string, error) {
+			ref, err := dockerArchiveTransport.ParseReference(path)
+			if err != nil {
+				return nil, err
+			}
+			return r.copyFromDockerArchive(ctx, ref, &options.CopyOptions)
+		},
+
+		// Give a decent error message if nothing above worked.
+		func() ([]string, error) {
+			return nil, errors.New("payload does not match any of the supported image formats (oci, oci-archive, dir, docker-archive)")
+		},
+	} {
+		loadedImages, loadError = f()
+		if loadError == nil {
+			return loadedImages, loadError
+		}
+	}
+
+	return nil, loadError
```

Comment:
This is always going to fail with  r.copyFromDockerArchive(ctx, ref, &options.CopyOptions) error, is this expected? Or should we have a generic error.

Reply:
We have a generic error in line 69:
```
 return nil, errors.New("payload does not match any of the supported image formats (oci, oci-archive, dir, docker-archive)")
```
We're hitting that if all previous attempts to load oci, docker archives etc. have failed.